mlir/lib/Conversion/GENToSPIRV/GENToSPIRV.cpp

```diff
@@ -22,13 +24,78 @@ namespace mlir {
 
 using namespace mlir;
 
+//===----------------------------------------------------------------------===//
+// ND-range Ops Lowerings
+//===----------------------------------------------------------------------===//
+
+/// Pattern to convert GEN3DNDRange operations to SPIR-V.
+///
+/// Convert:
+/// ```mlir
+/// %0 = gen.operation_name %dim
+///  ```
+/// To:
+/// ```mlir
+/// %__builtin__BuiltinName___addr = spirv.mlir.addressof
+/// @__builtin__BuiltinName__ : !spirv.ptr<vector<3xIndexType>, Input>
+/// %__builtin_value = spirv.Load "Input" %__builtin__BuiltinName___addr :
+/// vector<3xIndexType>
+/// %0 = spirv.VectorExtractDynamic %__builtin_value[%dim] :
+/// vector<3xIndexType>, i32
+/// ```
+/// With `__builtin__BuiltinName__` the name of a SPIR-V builtin, and
+/// `IndexType`, `i32` for 32-bit targets and `i64` for 64-bit targets.
+class GEN3DNDRangeLoweringBase : public ConversionPattern {
+public:
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    assert(operands.size() == 1 && "Expecting a single operand");
+    // The builtin variable must be of type <3xi32> for 32-bit targets and
+    // <3xi64> for 64-bit targets.
+    Type builtinType =
+        this->template getTypeConverter<SPIRVTypeConverter>()->getIndexType();
+    Value vector =
+        spirv::getBuiltinVariableValue(op, builtin, builtinType, rewriter);
+    rewriter.replaceOpWithNewOp<spirv::VectorExtractDynamicOp>(op, vector,
```

Comment:
Do we end up with more efficient binaries if we replace the dynamic extraction with a static extraction in case the input to the operation is a constant value?

Reply:
Very same LLVM representation: both will be translated to an LLVM `extractelement` instruction.